File: sdk/python/ecash_sdk.py
"""
Ecash Protocol — Solana Mining SDK (Python)
Full documentation: https://github.com/ecashprotocol/ecash-solana/blob/main/skill/SKILL.md
"""

import json
import re
from typing import Optional, List, Dict, Any
import requests

# ============ Constants ============
PROGRAM_ID = "w4eVWehdAiLdrxYduaF6UvSxCXTj2uAnstHJTgucwiY"
MINT = "7ePGWB6HaHhwucuBXuu4mVVGYryvibtWxPVYCgvtjRC7"
GLOBAL_STATE = "Bswa2hSMZKhN2MVMMFUSX9QqT7MPyUzfSnp2VyjmtUiS"
VAULT = "9nhEukfrhisGX1wu7gRmPGucZ76H1UC5mMPh8xhBgM7y"
API_URL = "https://api.ecash.bot"


# ============ Answer Normalization ============
def normalize(answer: str) -> str:
    """
    Normalize answer: lowercase, strip non-alphanumeric (except spaces), collapse spaces.

    Examples:
        >>> normalize("Hello,   World!")
        'hello world'
        >>> normalize("A.B.C. Test!!!")
        'abc test'
    """
    s = answer.lower()
    s = re.sub(r'[^a-z0-9 ]', '', s)
    s = re.sub(r'\s+', ' ', s).strip()
    return s


# ============ Blob Decryption ============
def decrypt_blob(puzzle_id: int, answer: str) -> Dict[str, Any]:
    """
    Decrypt puzzle blob using scrypt + AES-256-GCM.

    Args:
        puzzle_id: The puzzle ID (0-6299)
        answer: Your answer guess

    Returns:
        Dict with 'salt' and 'proof' if answer is correct

    Raises:
        ValueError: If answer is wrong (decryption fails)
        ImportError: If cryptography package not installed
    """
    try:
        from hashlib import scrypt as _scrypt
        from cryptography.hazmat.primitives.ciphers.aead import AESGCM
    except ImportError:
        raise ImportError("Install dependencies: pip install cryptography")

    normalized = normalize(answer)
    password = normalized.encode('utf-8')
    salt = f"ecash-v3-{puzzle_id}".encode('utf-8')

    # scrypt: N=131072, r=8, p=1, keyLen=32
    key = _scrypt(password, salt=salt, n=131072, r=8, p=1, dklen=32)

    # Fetch blob from API
    resp = requests.get(f"{API_URL}/puzzles/{puzzle_id}")
    puzzle = resp.json()

    if 'blob' not in puzzle or 'nonce' not in puzzle or 'tag' not in puzzle:
        raise ValueError(f"Puzzle {puzzle_id} blob data not found")

    nonce = bytes.fromhex(puzzle['nonce'])
    tag = bytes.fromhex(puzzle['tag'])
    ciphertext = bytes.fromhex(puzzle['blob'])

    # AES-256-GCM decrypt (ciphertext + tag)
    aesgcm = AESGCM(key)
    try:
        plaintext = aesgcm.decrypt(nonce, ciphertext + tag, None)
        return json.loads(plaintext.decode('utf-8'))
    except Exception as e:
        raise ValueError(f"Wrong answer (decryption failed): {e}")


# ============ Commit Hash ============
def compute_commit_hash(answer: str, salt_hex: str, secret_hex: str, signer_bytes: bytes) -> bytes:
    """
    Compute keccak256(answer || salt || secret || signer).

    Args:
        answer: The normalized answer
        salt_hex: Salt from decrypted blob (hex string, with or without 0x prefix)
        secret_hex: Random 32-byte secret (hex string)
        signer_bytes: Signer's public key (32 bytes)

    Returns:
        32-byte keccak256 hash
    """
    try:
        from sha3 import keccak_256
    except ImportError:
        try:
            from Crypto.Hash import keccak
            normalized = normalize(answer)
            combined = (
                normalized.encode('utf-8') +
                bytes.fromhex(salt_hex.replace('0x', '')) +
                bytes.fromhex(secret_hex.replace('0x', '')) +
                signer_bytes
            )
            k = keccak.new(digest_bits=256)
            k.update(combined)
            return k.digest()
        except ImportError:
            raise ImportError("Install sha3: pip install pysha3 or pycryptodome")

    normalized = normalize(answer)
    combined = (
        normalized.encode('utf-8') +
        bytes.fromhex(salt_hex.replace('0x', '')) +
        bytes.fromhex(secret_hex.replace('0x', '')) +
        signer_bytes
    )
    return keccak_256(combined).digest()


# ============ API Helpers ============
def get_puzzles(limit: int = 20, offset: int = 0, unsolved_only: bool = False) -> List[Dict]:
    """Get puzzles from the API."""
    params = {'limit': limit, 'offset': offset}
    if unsolved_only:
        params['unsolved'] = 'true'
    resp = requests.get(f"{API_URL}/puzzles", params=params)
    return resp.json()


def get_unsolved_puzzles(limit: int = 20) -> List[Dict]:
    """Get only unsolved puzzles."""
    puzzles = get_puzzles(limit=limit * 2)  # Fetch more to filter
    return [p for p in puzzles if not p.get('solved')][:limit]


def get_puzzle(puzzle_id: int) -> Dict:
    """Get a single puzzle by ID."""
    resp = requests.get(f"{API_URL}/puzzles/{puzzle_id}")
    return resp.json()


def get_stats() -> Dict:
    """Get protocol statistics."""
    resp = requests.get(f"{API_URL}/stats")
    return resp.json()


def get_leaderboard() -> List[Dict]:
    """Get the mining leaderboard."""
    resp = requests.get(f"{API_URL}/leaderboard")
    return resp.json()


def get_activity(limit: int = 10) -> List[Dict]:
    """Get recent solve activity."""
    resp = requests.get(f"{API_URL}/activity", params={'limit': limit})
    return resp.json()


def get_contract() -> Dict:
    """Get contract addresses and IDL."""
    resp = requests.get(f"{API_URL}/contract")
    return resp.json()


# ============ CLI ============
if __name__ == "__main__":
    import sys

    print("Ecash Protocol SDK")
    print(f"Program: {PROGRAM_ID}")
    print(f"Token: {MINT}")
    print()

    stats = get_stats()
    print(f"Total Solved: {stats.get('totalSolved', 'N/A')}")
    print(f"Current Batch: {stats.get('currentBatch', 'N/A')}")
    print(f"Current Era: {stats.get('currentEra', 'N/A')}")
    print(f"Reward per Solve: {stats.get('rewardPerPuzzle', 'N/A')} ECASH")
    print()

    unsolved = get_unsolved_puzzles(limit=5)
    print(f"Sample Unsolved Puzzles ({len(unsolved)} shown):")
    for p in unsolved[:3]:
        print(f"  #{p['puzzleId']}: {p['title']}")

File: sdk/js/ecash-sdk.mjs
/**
 * Ecash Protocol — Solana Mining SDK
 * Full documentation: https://github.com/ecashprotocol/ecash-solana/blob/main/skill/SKILL.md
 */

import { Connection, PublicKey, Keypair } from '@solana/web3.js';
import { TOKEN_2022_PROGRAM_ID, getAssociatedTokenAddressSync } from '@solana/spl-token';
import * as anchor from '@coral-xyz/anchor';
import crypto from 'crypto';

// ============ Constants ============
export const PROGRAM_ID = new PublicKey('w4eVWehdAiLdrxYduaF6UvSxCXTj2uAnstHJTgucwiY');
export const MINT = new PublicKey('7ePGWB6HaHhwucuBXuu4mVVGYryvibtWxPVYCgvtjRC7');
export const API_URL = 'https://api.ecash.bot';

// ============ PDA Derivation ============
export function getGlobalStatePda() {
  return PublicKey.findProgramAddressSync(
    [Buffer.from('global_state')],
    PROGRAM_ID
  )[0];
}

export function getMinerStatePda(owner) {
  return PublicKey.findProgramAddressSync(
    [Buffer.from('miner_state'), owner.toBuffer()],
    PROGRAM_ID
  )[0];
}

export function getVaultPda() {
  return PublicKey.findProgramAddressSync(
    [Buffer.from('vault')],
    PROGRAM_ID
  )[0];
}

export function getPuzzleSolvedPda(puzzleId) {
  const buf = Buffer.alloc(8);
  buf.writeBigUInt64LE(BigInt(puzzleId));
  return PublicKey.findProgramAddressSync(
    [Buffer.from('puzzle_solved'), buf],
    PROGRAM_ID
  )[0];
}

// ============ Answer Normalization ============
export function normalize(answer) {
  return answer
    .toLowerCase()
    .replace(/[^a-z0-9 ]/g, '')
    .replace(/\s+/g, ' ')
    .trim();
}

// ============ Blob Decryption ============
export async function decryptBlob(puzzleId, answer) {
  // Dynamically import scrypt-js
  const scryptPkg = await import('scrypt-js');
  const { scrypt } = scryptPkg.default || scryptPkg;

  const normalized = normalize(answer);
  const password = Buffer.from(normalized, 'utf-8');
  const salt = Buffer.from(`ecash-v3-${puzzleId}`, 'utf-8');

  // scrypt: N=131072, r=8, p=1, keyLen=32
  const keyArray = await scrypt(password, salt, 131072, 8, 1, 32);
  const key = Buffer.from(keyArray);

  // Fetch puzzle blob from API
  const res = await fetch(`${API_URL}/puzzles/${puzzleId}`);
  const puzzle = await res.json();

  if (!puzzle.blob || !puzzle.nonce || !puzzle.tag) {
    throw new Error('Puzzle blob data not found');
  }

  // AES-256-GCM decrypt
  const decipher = crypto.createDecipheriv(
    'aes-256-gcm',
    key,
    Buffer.from(puzzle.nonce, 'hex')
  );
  decipher.setAuthTag(Buffer.from(puzzle.tag, 'hex'));

  let decrypted = decipher.update(Buffer.from(puzzle.blob, 'hex'), null, 'utf-8');
  decrypted += decipher.final('utf-8');

  return JSON.parse(decrypted); // { salt, proof }
}

// ============ Commit Hash ============
export function computeCommitHash(answer, saltHex, secretHex, signerPubkey) {
  // Import keccak synchronously
  const { keccak_256 } = require('js-sha3');

  const normalized = normalize(answer);
  const answerBytes = Buffer.from(normalized, 'utf-8');
  const saltBytes = Buffer.from(saltHex.replace(/^0x/, ''), 'hex');
  const secretBytes = Buffer.from(secretHex.replace(/^0x/, ''), 'hex');
  const signerBytes = signerPubkey.toBuffer();

  const combined = Buffer.concat([answerBytes, saltBytes, secretBytes, signerBytes]);
  return Buffer.from(keccak_256.arrayBuffer(combined));
}

// ============ API Helpers ============
export async function getUnsolvedPuzzles() {
  const res = await fetch(`${API_URL}/puzzles`);
  const puzzles = await res.json();
  return puzzles.filter(p => !p.solved);
}

export async function getPuzzle(puzzleId) {
  const res = await fetch(`${API_URL}/puzzles/${puzzleId}`);
  return res.json();
}

export async function getStats() {
  const res = await fetch(`${API_URL}/stats`);
  return res.json();
}

export async function getLeaderboard() {
  const res = await fetch(`${API_URL}/leaderboard`);
  return res.json();
}

export async function getActivity(limit = 10) {
  const res = await fetch(`${API_URL}/activity?limit=${limit}`);
  return res.json();
}

export async function getContract() {
  const res = await fetch(`${API_URL}/contract`);
  return res.json();
}

// ============ Wallet Helpers ============
export function generateWallet() {
  const keypair = Keypair.generate();
  return {
    publicKey: keypair.publicKey.toBase58(),
    secretKey: Array.from(keypair.secretKey),
    keypair
  };
}

export function walletFromSecretKey(secretKeyArray) {
  const keypair = Keypair.fromSecretKey(new Uint8Array(secretKeyArray));
  return {
    publicKey: keypair.publicKey.toBase58(),
    keypair
  };
}

// ============ Token Account ============
export function getTokenAccount(owner) {
  return getAssociatedTokenAddressSync(
    MINT,
    owner,
    false,
    TOKEN_2022_PROGRAM_ID
  );
}
